### PR TITLE
Run clang tests with ASAN 

### LINKED ISSUE #5868: Build solc with asan for testing on CI
Goal is to detected memory leaks in the current test cases.

### PR DESCRIPTION
**Pull request by axic**
Closes #5868.

### CODE REVIEW

**Comment by axic** [comment]
Currently this fails because it finds issues. We could:
1) close this PR
2) ignore the failures, but merge it - with the hope that we'll fix them soon and not ignore the failures anymore
3) keep this open until they are fixed

File: .circleci/config.yml [hunk 1]
```diff
@@ -152,7 +152,7 @@ jobs:
       TERM: xterm
       CC: /usr/bin/clang-7
       CXX: /usr/bin/clang++-7
-      CMAKE_OPTIONS: -DLLL=ON
+      CMAKE_OPTIONS: -DSANITIZE=address -DCMAKE_BUILD_TYPE=Debug
```
Comment (bshastry) [inline .circleci/config.yml:175]: Relaying the argument of `-DSANITIZE` to 
https://github.com/ethereum/solidity/blob/develop/cmake/EthCompilerSettings.cmake#L141 may be a problem in two cases: (1) we want to add additional compiler flags per sanitizer; (2) if we want to the cmake extensible for adding more error detectors (e.g., undefined sanitizer) later on.

```
if (SANITIZE)
	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-omit-frame-pointer -fsanitize=${SANITIZE}")
endif()
```

For example, if `SANITIZER=address` we may want to add additional flags for debugging. And if `SANITIZER=undefined` we may want to use several undefined flags such as those listed here: https://github.com/google/oss-fuzz/blob/master/infra/base-images/base-builder/Dockerfile#L22

Comment (axic) [reply .circleci/config.yml:175]: Yes this is a known issue. Can you open an issue for it? (It is not related to this PR, but our CMake options).

Comment (bshastry) [reply .circleci/config.yml:175]: okay, filed here: #5934

**Review by chriseth — DISMISSED** [review]

**Comment by axic** [comment]
Even though in Github it is not considered a failure in terms of CI, yet the red mark is next to the PR so because of that it would be deterimental :(

Suggest to merge this, but commenting out trigger to run it.

**Comment by chriseth** [comment]
@axic ok! By the way, the memory leak seems to be connected to recursive structs, i.e. legal ones using mappings.

**Comment by chriseth** [comment]
This should always create an ASAN build, but only run the tests on the develop branch, not on pull requests.

**Review by ekpyron — APPROVED** [review]
Rebased and squashed it. Can be merged once tests pass again.
EDIT: or in fact can't, since the renamed test run was considered "required" :-).

**Pull request merged**